File: src/commands/show_rule.rs
use std::time::Duration;

use serenity::model::prelude::interaction::application_command::ApplicationCommandInteraction;
use serenity::prelude::*;
use serenity::{builder::CreateApplicationCommand, model::Permissions};

use crate::handler::snowflake_as_db;
use crate::DatabasePoolHolder;

use super::CommandResponse;

pub fn register(command: &mut CreateApplicationCommand) -> &mut CreateApplicationCommand {
    command
        .name("rule")
        .description("Show currently active rule in the server")
        .default_member_permissions(Permissions::ADMINISTRATOR)
        .dm_permission(false)
}

pub async fn run(
    ctx: &Context,
    command: &ApplicationCommandInteraction,
) -> Result<Option<CommandResponse>, String> {
    let type_map = ctx.data.read().await;
    let pool_holder: &DatabasePoolHolder = type_map.get::<DatabasePoolHolder>().unwrap();
    let pool = pool_holder.as_ref();

    let guild_id = snowflake_as_db(
        command
            .guild_id
            .expect("Received guild-only interaction without guild provided?!"),
    );

    let rule = match sqlx::query!("SELECT * FROM guild_rule WHERE guild_id = ?", guild_id)
        .fetch_one(pool)
        .await
    {
        Ok(rule) => rule,
        Err(why) => {
            return Err(format!(
                "Error while getting current active rule (did you `/setup`?): {why}"
            ))
        }
    };

    let mut text = vec![];
    text.push(format!("Current rule:"));
    text.push(format!(
        "System message channel: <#{}> `/syschannel`",
        rule.system_message_channel_id
    ));
    if let Some(c) = rule.pending_kick_notification_channel_id {
        text.push(format!(
            "Will remind about pending kicks in: <#{}> `/pendingchannel`",
            c
        ));
    } else {
        text.push(format!(
            "Will remind about pending kicks in: unset `/pendingchannel`"
        ));
    }
    text.push(format!(
        "Reminders will be sent at seconds before kick: {:?} `/pendingreminders`",
        rule.pending_kick_notification_values
    ));

    if let Some(t) = rule.new_member_kick_timeout {
        let dur = Duration::from_secs(t as u64);
        text.push(format!(
            "New members get kicked after: {} seconds = {} `/newtimeout`",
            t,
            humantime::format_duration(dur)
        ));
    } else {
        text.push(format!("New members get kicked after: never `/newtimeout`"));
    }

    if let Some(r) = rule.immunity_role_id {
        text.push(format!("Kick immunity role: <@&{}> `/immunityrole`", r));
    } else {
        text.push(format!("Kick immunity role: unset `/immunityrole`"))
    }

    if let Some(t) = rule.loss_of_immunity_role_timeout {
        let dur = Duration::from_secs(t as u64);
        text.push(format!(
            "Members losing immunity role get kicked after: {} seconds = {} `/immunitytimeout`",
            t,
            humantime::format_duration(dur)
        ));
    } else {
        text.push(format!(
            "Members losing immunity role get kicked after: never `/immunitytimeout`"
        ));
    }

    let dur = Duration::from_secs(rule.kick_safety_timeout as u64);
    text.push(format!(
        "When kicking a member, allow cancelling within: {} seconds = {} `/safetytimeout`",
        rule.kick_safety_timeout,
        humantime::format_duration(dur)
    ));

    Ok(CommandResponse::text(text.join("\n")))
}

File: commands/show_rule.py
import discord
from database import *
from discord import app_commands
import humanize
import datetime as dt

@app_commands.guild_only()
async def show_config(interaction: discord.Interaction):
    """Display the current configuration of this server."""
    guild_id = interaction.guild_id
    try:
        config = GuildConfig.get(GuildConfig.guild_id == guild_id)
    except pw.DoesNotExist:
        await interaction.response.send_message("No config found for the current server. Run `/setup` to perform initial setup.")
        return

    text = 'Current server config:\n'
    text += f"System message channel: <#{config.system_message_channel_id}> `/syschannel`\n"

    if config.pending_kick_notification_channel_id:
        text += f"Will send reminders about pending kicks in: <#{config.pending_kick_notification_channel_id}> `/pendingchannel`\n"
    else:
        text += f"Will send reminders about pending kicks in: unset `/pendingchannel`\n"

    text += f'When the time until kick passes these thresholds, send a reminder: "{config.pending_kick_notification_values}" `/pendingreminders`\n'

    if config.new_member_kick_timeout:
        h = humanize.precisedelta(dt.timedelta(seconds=config.new_member_kick_timeout))
        text += f'New members get kicked after: {h} = {config.new_member_kick_timeout} seconds `/newtimeout`\n'
    else:
        text += 'New members get kicked after: never `/newtimeout`\n'

    if config.immunity_role_id:
        text += f'Kick immunity role: <@&{config.immunity_role_id}> `/immunityrole`\n'
    else:
        text += 'Kick immunity role: unset `/immunityrole`\n'
    
    if config.loss_of_immunity_role_timeout:
        h = humanize.precisedelta(dt.timedelta(seconds=config.loss_of_immunity_role_timeout))
        text += f'Members losing immunity role get kicked after: {h} = {config.loss_of_immunity_role_timeout} seconds `/immunitytimeout`\n'
    else:
        text += 'Members losing immunity role get kicked after: never `/immunitytimeout`\n'

    h = humanize.precisedelta(dt.timedelta(seconds=config.kick_safety_timeout))
    text += f'When kicking a member, allow cancelling within: {h} = {config.kick_safety_timeout} seconds `/safetytimeout`\n'

    text += 'To reset these settings to safe defaults, perform `/setup` again.'
    await interaction.response.send_message(text)


show_config = app_commands.Command(
    name="showconfig",
    description="Display the current configuration of this server.",
    callback=show_config
)
show_config.default_permissions = discord.Permissions()
show_config.default_permissions.kick_members = True
